tests/unit/steps/test_steps_plan.py

```diff
@@ -441,3 +451,156 @@ def test_analysis_not_print_warn_manually_upgrade(mock_print, model, apps):
     )
     manually_upgrade_data_plane(result)
     mock_print.assert_not_called()
+
+
+@pytest.mark.parametrize(
+    "machines, hostnames, azs",
+    [
+        ({"1", "2", "3"}, None, None),
+        (None, {"juju-cef358-1"}, None),
+        (None, None, {"zone-1"}),
+        (None, None, None),
+    ],
+)
+@patch("cou.steps.plan.verify_data_plane_cli_azs")
+@patch("cou.steps.plan.verify_data_plane_cli_hostnames")
+@patch("cou.steps.plan.verify_data_plane_cli_machines")
+def test_verify_data_plane_cli_input(
+    mock_verify_machines,
+    mock_verify_hostnames,
+    mock_verify_azs,
+    machines,
+    hostnames,
+    azs,
+    model,
+    apps,
+    cli_args,
+):
+    result = Analysis(
+        model=model,
+        apps_control_plane=[apps["keystone_focal_ussuri"]],
+        apps_data_plane=[apps["nova_focal_ussuri"]],
+    )
+    cli_args.machines = machines
+    cli_args.hostnames = hostnames
+    cli_args.availability_zones = azs
+
+    verify_data_plane_cli_input(cli_args, result)
+
+    if machines:
+        mock_verify_machines.assert_called_once()
+        mock_verify_hostnames.assert_not_called()
+        mock_verify_azs.assert_not_called()
+
+    if hostnames:
+        mock_verify_machines.assert_not_called()
+        mock_verify_hostnames.assert_called_once()
+        mock_verify_azs.assert_not_called()
+
+    if azs:
+        mock_verify_machines.assert_not_called()
+        mock_verify_hostnames.assert_not_called()
+        mock_verify_azs.assert_called_once()
```

Comment:
non-blocking: If we need to have this many conditions, it would be better separating them into different functions instead of forcing them into one parametrized test function.